### PR TITLE
Feature 2708 mvmode merge flag default

### LINKED ISSUE #2708: Enhance Multivariate MODE to change the default `merge_flag` setting to NONE
## Describe the Enhancement ##
Change the `merge_flag default` to` NONE`, and allow it to not be set for MvMode fields. If not set right now the code uses a default `merge_flag=THRESH` and `merge_thresh = ">=1.25"`. MvMode is now implemented so that the user MUST set all the merge_flag (and merge_thresh) values explicitly. 

### Time Estimate ###
1 day

### Sub-Issues ###
Consider breaking the enhancement down into sub-issues.
- [ ] See if changing for both mvmode and single var mode causes problems for single var mode.

### Relevant Deadlines ###
November 7 2023 (beta 2)

### Funding Source ###
2702691

### Assignee ###
- [x] Select **engineer(s)** or **no engineer** required
- [x] Select **scientist(s)** or **no scientist** required

### Labels ###
- [ ] Review default **alert** labels
- [ ] Select **component(s)**
- [ ] Select **priority**
- [ ] Select **requestor(s)**

### Milestone and Projects ###
- [ ] Select **Milestone** as the next official version or **Backlog of Development Ideas**
- [ ] For the next official version, select the **MET-X.Y.Z Development** project

## Define Related Issue(s) ##
Consider the impact to the other METplus components.
- [ ] [METplus](https://github.com/dtcenter/METplus/issues/new/choose), [MET](https://github.com/dtcenter/MET/issues/new/choose)

## Enhancement Checklist ##
See the [METplus Workflow](https://metplus.readthedocs.io/en/latest/Contributors_Guide/github_workflow.html) for details.
- [x] Complete the issue definition above, including the **Time Estimate** and **Funding Source**.
- [x] Fork this repository or create a branch of **develop**.
Branch name: `feature_<Issue Number>_<Description>`
- [x] Complete the development and test your changes.
- [x] Add/update log messages for easier debugging.
- [x] Add/update unit tests.
- [x] Add/update documentation.
- [x] Push local changes to GitHub.
- [x] Submit a pull request to merge into **develop**.
Pull request: `feature <Issue Number> <Description>`
- [x] Define the pull request metadata, as permissions allow.
Select: **Reviewer(s)** and **Development** issue
Select: **Milestone** as the next official version
Select: **MET-X.Y.Z Development** project for development toward the next official release
- [x] Iterate until the reviewer(s) accept and merge your changes.
- [x] Delete your fork or branch.
- [ ] Close this issue.

### PR DESCRIPTION
**Pull request by davidalbo**
## Expected Differences ##

- [x] Do these changes introduce new tools, command line arguments, or configuration file options? **[No]**</br>

- [x] Do these changes modify the structure of existing or add new output data types (e.g. statistic line types or NetCDF variables)? **[No]**</br>

## Pull Request Testing ##

- [x] Describe testing already performed for these changes:</br>  I did these tests:

1. create a config with a field that does not specify merge_flag/merge_thresh and make sure it defaults to NONE.
2. create a config with a field that has merge_flag but no merge_thresh and observe the error exit.
3. create a config with a field that has merge_thresh but no merge_flag and observe the error exit.
</br>

- [x] Recommend testing for the reviewer(s) to perform, including the location of input datasets, and any additional instructions:</br>  Take a previous test case (if still around) and modify the config file in the above 3 ways to see what happens.

- [x] Do these changes include sufficient documentation updates, ensuring that no errors or warnings exist in the build of the documentation? **[Yes]**

- [x] Do these changes include sufficient testing updates? **[Yes ]**

- [x] Will this PR result in changes to the test suite? **[No]**</br>
If **yes**, describe the new output and/or changes to the existing output:</br>

- [x] Please complete this pull request review by **[October 17 2023]**.</br>

## Pull Request Checklist ##
See the [METplus Workflow](https://metplus.readthedocs.io/en/latest/Contributors_Guide/github_workflow.html) for details.
- [x] Review the source issue metadata (required labels, projects, and milestone).
- [x] Complete the PR definition above.
- [x] Ensure the PR title matches the feature or bugfix branch name.
- [ ] Define the PR metadata, as permissions allow.
Select: **Reviewer(s)** and **Development** issue
Select: **Milestone** as the version that will include these changes
Select: **Coordinated METplus-X.Y Support** project for bugfix releases or **MET-X.Y.Z Development** project for official releases
- [x] After submitting the PR, select the :gear: icon in the **Development** section of the right hand sidebar. Search for the issue that this PR will close and select it, if it is not already selected.
- [x] After the PR is approved, merge your changes. If permissions do not allow this, request that the reviewer do the merge.
- [x] Close the linked issue and delete your feature or bugfix branch from GitHub.

### CODE REVIEW

**Comment by hertneky** [comment]
@davidalbo Should I be testing this also? If so, where is it installed? Or am I just reviewing the code changes?

**Comment by davidalbo** [comment]
@hertneky I put a fresh mode executable built from this branch to /scratch/dave
If you could do some tests by running that, that'd be what I'm hoping for.  (recommended tests above).

That plus look at the documentation changes.     

I'd leave it to @JohnHalleyGotway  to look at code changes.

File: src/libcode/vx_shapedata/mode_field_info.cc [hunk 1]
```diff
@@ -219,32 +221,87 @@ if ( dict->lookup(conf_key_conv_thresh) )  {
...
+               string thresh_str = merge_thresh_array.thresh()->get_str();
+               mlog  << Debug(2) << "Field:" << name << ". Using parent merge_flag: "
+                     << merge_name << " and parent merge_thresh:" << thresh_str
+                     << "\n";
+            }
+         }
+      }
    }
-} else {
+ } else {
```
Comment (JohnHalleyGotway) [inline src/libcode/vx_shapedata/mode_field_info.cc]:
```suggestion
} else {
```
Minor point. Just aligning braces.

File: src/libcode/vx_shapedata/mode_field_info.cc [hunk 2]
```diff
@@ -219,32 +221,87 @@ if ( dict->lookup(conf_key_conv_thresh) )  {
...
+      string merge_name = mergetype_to_string(merge_flag);
+      if (dict->lookup(conf_key_merge_thresh, false)) {
+         // the individual entry also has a merge_thresh, this is good
+         merge_thresh_array = dict->lookup_thresh_array(conf_key_merge_thresh, 
+                                                        default_dictionary_error_out,
+                                                        default_dictionary_print_warning,
+                                                        false);
+      } else {
+         // get the parent's merge_thresh, just to have something.  Error out if the merge_flag is not none
+         // becAuse that is inconsistent
```
Comment (JohnHalleyGotway) [inline src/libcode/vx_shapedata/mode_field_info.cc]:
```suggestion
         // because that is inconsistent
```

**Review by JohnHalleyGotway — APPROVED** [review]
I approve of these changes... with a couple minor proposed fixes to spacing and capitalization.

I note that all the GHA tests all ran and flagged no diffs. I reviewed the code changes, was able to step through the logic described in the comments, and think it makes sense. I also see that MODE is now making use of the default `MODEMultivarConfig_default` file that already existed but wasn't ever being read. Thanks for updating the user's guide to reference that config file.

The User's Guide *could* be updated to specifically outline this logic for how the `merge_flag` and `merge_thresh` config options are parsed. But I suspect that's too much detail for the User's Guide. If @hertneky sees it differently and would like this logic described there, then I defer to her.

Thanks for getting this one across the finish line @davidalbo!

**Comment by hertneky** [comment]
Test cntrl run with merge flag=THRESH and thresholds specified for the various fields. -> Object for fcst/obs look okay, all objects are merged into 1 cluster for both fcst/obs.

Test not specifying merge_flag/merge_thresh and make sure it defaults to NONE. -> Defaults to None and uses >=3.5 for all vars. Not sure why it specifies a random value for thresh, when merge is set to None, but it doesn't matter I suppose, since it is not used. Objects look the same as cntrl and no merging/clustered objects as expected.

Test specifying merge_flag=THRESH but no merge_thresh and observe the error exit. -> Errors as expected - need to specify a threshold.

Test usage of merge_thresh but no merge_flag and observe the error exit. -> Errors as expected - need to specify merge_flag

**Review by hertneky — APPROVED** [review]

**Pull request merged**